### ISSUE TITLE
High virtual memory usage on macOS

### ISSUE THREAD

**Issue by roberthharvey** [citra-emu/citra#2416]
issue 1 https://github.com/citra-emu/citra/issues/2415

here is the ram problem (it keeps building... did not wait till it got to breaking point this time only to about 30GBs)

<img width="510" alt="screen shot 2017-01-08 at 3 44 54 pm" src="https://cloud.githubusercontent.com/assets/12015507/21748198/82f3e05e-d5b9-11e6-89c3-ab502ae9540e.png">
<img width="984" alt="screen shot 2017-01-08 at 3 20 03 pm" src="https://cloud.githubusercontent.com/assets/12015507/21748200/8342559a-d5b9-11e6-81b9-4eb781faa79c.png">

game: PKM Moon, build (tried last 4, same issue)

Hardware Overview:

Model Name:	MacBook Pro
Model Identifier:	MacBookPro11,4
Processor Name:	Intel Core i7
Processor Speed:	2.2 GHz
Number of Processors:	1
Total Number of Cores:	4
L2 Cache (per Core):	256 KB
L3 Cache:	6 MB
Memory:	16 GB
Boot ROM Version:	MBP114.0172.B10
SMC Version (system):	2.29f24

**Comment by roberthharvey** [comment]
https://www.dropbox.com/s/eiye8t4o7qik4qs/Terminal%20Saved%20Output.TXT?dl=0

link is the dump file

<img width="564" alt="screen shot 2017-01-08 at 4 35 49 pm" src="https://cloud.githubusercontent.com/assets/12015507/21748517/e189a09c-d5c1-11e6-8dc1-75812224555a.png">
<img width="1420" alt="screen shot 2017-01-08 at 4 29 31 pm" src="https://cloud.githubusercontent.com/assets/12015507/21748518/e1da04e2-d5c1-11e6-87df-f94cfade553e.png">

**Comment by mailwl** [comment]
never saw memory leaks on mac. it's pokemon only issue?

**Comment by cherrytwizzler** [comment]
My mac has the same issue with pokemon sun and virtue's last reward 
I don't have my computer on me rn but I know it has 8G ram and it's running on El capitan ver. 10.11.5 
My game gets the device has run out of application memory message at about 20G of memory usage on activity monitor though but I guess that's because you have more ram than I do

**Comment by roberthharvey** [comment]
there seams to be a tone of these 'StoreDataCache' errors because of unimplemented function? 



[  23.196817] Service <Error> core/hle/service/service.cpp:HandleSyncRequest:96: unknown / unimplemented function 'StoreDataCache': port=gsp::Gpu, cmd_buff[1]=0x302B3C00, cmd_buff[2]=0x20, cmd_buff[3]=0x0, cmd_buff[4]=0xFFFF8001
[  23.196875] Service <Error> core/hle/service/service.cpp:HandleSyncRequest:96: unknown / unimplemented function 'StoreDataCache': port=gsp::Gpu, cmd_buff[1]=0x302DF680, cmd_buff[2]=0x200, cmd_buff[3]=0x0, cmd_buff[4]=0xFFFF8001
[  23.197031] Service <Error> core/hle/service/service.cpp:HandleSyncRequest:96: unknown / unimplemented function 'StoreDataCache': port=gsp::Gpu, cmd_buff[1]=0x302D1680, cmd_buff[2]=0x1000, cmd_buff[3]=0x0, cmd_buff[4]=0xFFFF8001
[  23.197063] Service <Error> core/hle/service/service.cpp:HandleSyncRequest:96: unknown / unimplemented function 'StoreDataCache': port=gsp::Gpu, cmd_buff[1]=0x302DB200, cmd_buff[2]=0x80, cmd_buff[3]=0x0, cmd_buff[4]=0xFFFF8001
[  23.197144] Service <Error> core/hle/service/service.cpp:HandleSyncRequest:96: unknown / unimplemented function 'StoreDataCache': port=gsp::Gpu, cmd_buff[1]=0x302B3C80, cmd_buff[2]=0x100, cmd_buff[3]=0x0, cmd_buff[4]=0xFFFF8001
[  23.197186] Service <Error> core/hle/service/service.cpp:HandleSyncRequest:96: unknown / unimplemented function 'StoreDataCache': port=gsp::Gpu, cmd_buff[1]=0x302CE880, cmd_buff[2]=0x40, cmd_buff[3]=0x0, cmd_buff[4]=0xFFFF8001
[  23.197214] Service <Error> core/hle/service/service.cpp:HandleSyncRequest:96: unknown / unimplemented function 'StoreDataCache': port=gsp::Gpu, cmd_buff[1]=0x302A3500, cmd_buff[2]=0x10000, cmd_buff[3]=0x0, cmd_buff[4]=0xFFFF8001

then this repeats over and over and over and I think is the main memory leak problem

[ 898.092319] Service.NDM <Warning> core/hle/service/ndm/ndm.cpp:EnterExclusiveState:35: (STUBBED) exclusive_state=0x00000002 
[ 898.092352] Kernel.SVC <Warning> core/hle/svc.cpp:CreateMemoryBlock:927: called addr=0x083E5000
[ 898.092382] Service.NWM <Warning> core/hle/service/nwm/nwm_uds.cpp:InitializeWithVersion:115: (STUBBED) called unk1=0x00014000, unk2=0x00000400, value=0, handle=0x0014883E
[ 898.092395] Service.NDM <Warning> core/hle/service/ndm/ndm.cpp:LeaveExclusiveState:44: (STUBBED) exclusive_state=0x00000000 
[ 898.092417] Service.FRD <Warning> core/hle/service/frd/frd.cpp:GetMyFriendKey:91: (STUBBED) called
[ 898.115912] Service.NDM <Warning> core/hle/service/ndm/ndm.cpp:EnterExclusiveState:35: (STUBBED) exclusive_state=0x00000002 
[ 898.115954] Kernel.SVC <Warning> core/hle/svc.cpp:CreateMemoryBlock:927: called addr=0x083E5000
[ 898.116037] Service.NWM <Warning> core/hle/service/nwm/nwm_uds.cpp:InitializeWithVersion:115: (STUBBED) called unk1=0x00014000, unk2=0x00000400, value=0, handle=0x00148842
[ 898.116054] Service.NDM <Warning> core/hle/service/ndm/ndm.cpp:LeaveExclusiveState:44: (STUBBED) exclusive_state=0x00000000 
[ 898.116077] Service.FRD <Warning> core/hle/service/frd/frd.cpp:GetMyFriendKey:91: (STUBBED) called
[ 898.156073] Service.NDM <Warning> core/hle/service/ndm/ndm.cpp:EnterExclusiveState:35: (STUBBED) exclusive_state=0x00000002 
[ 898.156119] Kernel.SVC <Warning> core/hle/svc.cpp:CreateMemoryBlock:927: called addr=0x083E5000
[ 898.156154] Service.NWM <Warning> core/hle/service/nwm/nwm_uds.cpp:InitializeWithVersion:115: (STUBBED) called unk1=0x00014000, unk2=0x00000400, value=0, handle=0x00148846
[ 898.156167] Service.NDM <Warning> core/hle/service/ndm/ndm.cpp:LeaveExclusiveState:44: (STUBBED) exclusive_state=0x00000000 
[ 898.156190] Service.FRD <Warning> core/hle/service/frd/frd.cpp:GetMyFriendKey:91: (STUBBED) called
[ 898.195172] Service.NDM <Warning> core/hle/service/ndm/ndm.cpp:EnterExclusiveState:35: (STUBBED) exclusive_state=0x00000002 
[ 898.195218] Kernel.SVC <Warning> core/hle/svc.cpp:CreateMemoryBlock:927: called addr=0x083E5000
[ 898.195245] Service.NWM <Warning> core/hle/service/nwm/nwm_uds.cpp:InitializeWithVersion:115: (STUBBED) called unk1=0x00014000, unk2=0x00000400, value=0, handle=0x0014884A
[ 898.195259] Service.NDM <Warning> core/hle/service/ndm/ndm.cpp:LeaveExclusiveState:44: (STUBBED) exclusive_state=0x00000000 
[ 898.195281] Service.FRD <Warning> core/hle/service/frd/frd.cpp:GetMyFriendKey:91: (STUBBED) called

**Comment by mailwl** [comment]
no. unimplemented functions does nothing and don't waste resourses

**Comment by roberthharvey** [comment]
@mailwl ok, what about the: 

```
[5943.089491] Service.NDM <Warning> core/hle/service/ndm/ndm.cpp:EnterExclusiveState:35: (STUBBED) exclusive_state=0x00000002 
[5943.089521] Kernel.SVC <Warning> core/hle/svc.cpp:CreateMemoryBlock:927: called addr=0x083E5000
[5943.089549] Service.NWM <Warning> core/hle/service/nwm/nwm_uds.cpp:InitializeWithVersion:115: (STUBBED) called unk1=0x00014000, unk2=0x00000400, value=0, handle=0x0014A2E5
[5943.089561] Service.NDM <Warning> core/hle/service/ndm/ndm.cpp:LeaveExclusiveState:44: (STUBBED) exclusive_state=0x00000000 
[5943.089583] Service.FRD <Warning> core/hle/service/frd/frd.cpp:GetMyFriendKey:91: (STUBBED) called
```

it says, entering exclusive state. then create memory bock, then something about InitializeWithVersion and leaving this state and then GetMyFriendKey:91

I'm guessing this is the leak right? (I have absolutely no idea haha just trying to be helpful) :)

**Comment by freiro** [comment]
@roberthharvey those are stubbed functions and probably not the cause of the leak (afaik), let's wait for a developer with macOS.

**Comment by roberthharvey** [comment]
@freiro okidoki 

found these also

[citra-qt-bin_2017-01-08-122010_Roberts-MacBook-Pro.crash.TXT](https://github.com/citra-emu/citra/files/692214/citra-qt-bin_2017-01-08-122010_Roberts-MacBook-Pro.crash.TXT)
[citra-qt-bin_2017-01-08-143705_Roberts-MacBook-Pro.crash.TXT](https://github.com/citra-emu/citra/files/692215/citra-qt-bin_2017-01-08-143705_Roberts-MacBook-Pro.crash.TXT)

**Comment by roberthharvey** [comment]
found these too

[citra-qt_2017-01-08-180540_Roberts-MacBook-Pro.cpu_resource.diag.TXT](https://github.com/citra-emu/citra/files/692218/citra-qt_2017-01-08-180540_Roberts-MacBook-Pro.cpu_resource.diag.TXT)
[citra-qt_2017-01-08-180642_Roberts-MacBook-Pro.wakeups_resource.diag.TXT](https://github.com/citra-emu/citra/files/692223/citra-qt_2017-01-08-180642_Roberts-MacBook-Pro.wakeups_resource.diag.TXT)
[citra-qt_2017-01-08-182806_Roberts-MacBook-Pro.cpu_resource.diag.TXT](https://github.com/citra-emu/citra/files/692222/citra-qt_2017-01-08-182806_Roberts-MacBook-Pro.cpu_resource.diag.TXT)
[citra-qt_2017-01-08-183035_Roberts-MacBook-Pro.wakeups_resource.diag.TXT](https://github.com/citra-emu/citra/files/692220/citra-qt_2017-01-08-183035_Roberts-MacBook-Pro.wakeups_resource.diag.TXT)
[citra-qt_2017-01-08-221212_Roberts-MacBook-Pro.cpu_resource.diag.diag.TXT](https://github.com/citra-emu/citra/files/692224/citra-qt_2017-01-08-221212_Roberts-MacBook-Pro.cpu_resource.diag.diag.TXT)
[citra-qt_2017-01-08-221327_Roberts-MacBook-Pro.wakeups_resource.diag.TXT](https://github.com/citra-emu/citra/files/692221/citra-qt_2017-01-08-221327_Roberts-MacBook-Pro.wakeups_resource.diag.TXT)
[citra-qt-bin_2017-01-08-115322_Roberts-MacBook-Pro.wakeups_resource.diag.TXT](https://github.com/citra-emu/citra/files/692219/citra-qt-bin_2017-01-08-115322_Roberts-MacBook-Pro.wakeups_resource.diag.TXT)
[citra-qt-bin_2017-01-08-120119_Roberts-MacBook-Pro.wakeups_resource.diag.TXT](https://github.com/citra-emu/citra/files/692225/citra-qt-bin_2017-01-08-120119_Roberts-MacBook-Pro.wakeups_resource.diag.TXT)
[citra-qt-bin_2017-01-08-122240_Roberts-MacBook-Pro.wakeups_resource.diag.TXT](https://github.com/citra-emu/citra/files/692226/citra-qt-bin_2017-01-08-122240_Roberts-MacBook-Pro.wakeups_resource.diag.TXT)
[citra-qt-bin_2017-01-08-122325_Roberts-MacBook-Pro.cpu_resource.diag.TXT](https://github.com/citra-emu/citra/files/692227/citra-qt-bin_2017-01-08-122325_Roberts-MacBook-Pro.cpu_resource.diag.TXT)
[citra-qt-bin_2017-01-08-145208_Roberts-MacBook-Pro.wakeups_resource.diag.TXT](https://github.com/citra-emu/citra/files/692228/citra-qt-bin_2017-01-08-145208_Roberts-MacBook-Pro.wakeups_resource.diag.TXT)
[citra-qt-bin_2017-01-08-175025_Roberts-MacBook-Pro.wakeups_resource.diag.TXT](https://github.com/citra-emu/citra/files/692229/citra-qt-bin_2017-01-08-175025_Roberts-MacBook-Pro.wakeups_resource.diag.TXT)

**Comment by pippo2931** [comment]
See #1701 too.
There is a bunch of other games that have the same problem:
Paper Mario Sticker Stars, Yokai Watch, Shin Megamy Tensei 4
In top you can see CMPRS memory building up.

One way to slow the leak is to disable "scaled resolution"

**Comment by wwylele** [comment]
I am not a macOS user. If possible, could someone using macOS debug citra, and see if this memory is  "leaked"(meaning no live reference to them), or actually be occupied by some object? (for example a big vector or map) Which object is it? 

We used to have a similar issue, which turns out to be an `std::unordered_map` being inflated by items whose hash are wrongly calculated. Probably we can think in this way.

**Comment by roberthharvey** [comment]
@wwylele how? which debugger...

**Comment by mailwl** [comment]
Xcode is good

**Comment by roberthharvey** [comment]
@wwylele umm... how do you debug a live program in Xcode?...

**Comment by FernandoS27** [comment]
It's quite probably a problem with OpenGL Drivers in Mac.

http://aras-p.info/blog/2007/07/14/debugging-story-video-memory-leaks/

a fast fix would be forcing an unbinding of the textures when flushing the cache.

**Comment by Subv** [comment]
@roberthharvey Could you please try to use an OpenGL memory profiler to see if the GPU textures are what's causing the leak? Please report the video memory usage when playing the affected game.

**Comment by FernandoS27** [comment]
add this:
GLuint tmp[1] = {texture_to_flush}; 
glDeleteTextures(1,tmp); 

to: https://github.com/citra-emu/citra/blob/master/src/video_core/renderer_opengl/gl_rasterizer_cache.cpp#L756

if it reduces the leaks then that's what's happening.

**Comment by roberthharvey** [comment]
@FernandoS27 still goes mental
<img width="1436" alt="screen shot 2017-01-09 at 3 27 10 am" src="https://cloud.githubusercontent.com/assets/12015507/21752894/9c78b132-d61b-11e6-9b3c-83e1968005f8.png">

**Comment by jroweboy** [comment]
@roberthharvey that bit of code wasn't meant to fix the terminal spam (which is harmless). what about the ram usage?

**Comment by roberthharvey** [comment]
@jroweboy I am running the game now and will see, it takes a bit to ramp up

**Comment by roberthharvey** [comment]
@Subv this?

<img width="1320" alt="screen shot 2017-01-09 at 3 31 28 am" src="https://cloud.githubusercontent.com/assets/12015507/21752918/5d9ce82e-d61c-11e6-9ef6-b599ded563cd.png">
[Screen Recording 2017-01-09 at 3.30.46 AM.mov.zip](https://github.com/citra-emu/citra/files/692398/Screen.Recording.2017-01-09.at.3.30.46.AM.mov.zip)

@jroweboy I think its worse... ramped up really quickly...

[Sample of Citra.txt](https://github.com/citra-emu/citra/files/692401/Sample.of.Citra.txt)
<img width="461" alt="screen shot 2017-01-09 at 3 35 07 am" src="https://cloud.githubusercontent.com/assets/12015507/21752935/a8d83b86-d61c-11e6-99e3-dd006914cbbc.png">

**Comment by pippo2931** [comment]
Still leaking.
One thing I noticed that reduces the leak is using the native resolution.

**Comment by roberthharvey** [comment]
@pippo2931 I found dropping to the lowest was the best. 

@Subv opening the graphics debugger make the app unesponsive but it does not crash or pause emulation.

<img width="233" alt="screen shot 2017-01-09 at 3 41 06 am" src="https://cloud.githubusercontent.com/assets/12015507/21753003/d31d8c1a-d61d-11e6-91e9-bafef5bb3cf0.png">

**Comment by Lord-Kamina** [comment]
Just chiming in to say I was just running a version I compiled myself with pokemon sun on a hackintosh with an i7-3770 and a GTX980 and virtual memory didn't go above 3GB. Oddly enough though, I think it runs better on my actual 2011 MBP (Sandy bridge i7 can't remember exact model and Radeon 6770)

**Comment by FernandoS27** [comment]
the line I told to add, compare before adding the line and after adding it. Leaks won't be done but they should reduce significantly (there's still FBO leaking...)

**Comment by pippo2931** [comment]
I don't see a big change
After two minutes, 4X resolution, same scene, same game, 1 GB leaked. 
Original
PID    COMMAND      %CPU TIME     #TH  #WQ  #POR **MEM**   PURG  **CMPRS**  PGRP  PPID
47756  citra-qt     90.2 02:00.58 14/1 9    279  **481M+** 0B    **1023M+** 47756 38372 
With line change
PID    COMMAND      %CPU TIME     #TH  #WQ  #POR **MEM**   PURG **CMPRS**  PGRP  PPID
7011  citra-qt     85.0 02:03.43 14/1 9    294  **489M**+ 0B   **1008M+** 47011 38372

I assume that the compressed memory is the leaked memory because when compressed reaches 20GB my system crashes.

Native resolution has a bigger active memory usage, but a slower leakage to compressed.
PID    COMMAND      %CPU  TIME     #TH  #WQ  #POR **MEM**    PURG **CMPRS**  PGRP  PPID
48533  citra-qt     98.1  05:06.83 14/1 9    276  **3373M+** 0B   **214M**   48533 38372

**Comment by archshift** [comment]
@pippo2931 have you had the same issue with any games besides Pokemon Moon?

**Comment by archshift** [comment]
Never mind, it's happening on the Yo-Kai demo.

**Comment by FernandoS27** [comment]
Well in that case let's make sure where the issue is.

Run citra without hardware renderer, if the problem persists, run it without shader jit, if the problem persists run it without cpu jit.

**Comment by pippo2931** [comment]
Run citra without hardware renderer: leak
run it without shader jit: no leak - Only 263MB in use

**Comment by archshift** [comment]
Just for the record, running the Yo-Kai demo:

![](https://i.imgur.com/9oVd3CV.png)

**Comment by cherrytwizzler** [comment]
I tried running the game w/out Shader JIT but after the first five minutes the activity monitor spiked the same way as it does w/ JIT (I didn't wait for it to stop completely though). I also noticed the emulation slowed down considerably 
<img width="1256" alt="screen shot 2017-01-08 at 10 30 16 pm" src="https://cloud.githubusercontent.com/assets/24989213/21756691/7f81c2b8-d5f2-11e6-91be-822453e0140a.png">

**Comment by cherrytwizzler** [comment]
I tried running w/out the shader but the same thing happened after a few minutes (and the emulation was also very slow compared to running it with the shader)
<img width="1275" alt="screen shot 2017-01-08 at 10 43 50 pm" src="https://cloud.githubusercontent.com/assets/24989213/21756830/473e4ad2-d5f4-11e6-81ae-be5a5e6316a1.png">

**Comment by cherrytwizzler** [comment]
and running without the cpu jit does impede the leakage for a little before falling back into the same trend

**Comment by FernandoS27** [comment]
Well I'll check this out next once I find my doodle for my hackintosh.

**Comment by pippo2931** [comment]
Add Zelda Majoras Mask to the list.

**Comment by bouch44** [comment]
Has anyone found a solution to this? About a month later and no posts =/ Hoping so, seems like I can only run for about 10-15 minutes before Application Memory message returns. Hopefully have saved at that point.... Any help would be greatly appreciated =]

**Comment by roberthharvey** [comment]
so, I have not ran this app since this issue, no point. but it has been some time now so devs any news? 

anyone find a fix/workaround

**Comment by cherrytwizzler** [comment]
@roberthharvey the most recent builds of citra fixed the issue I was having with pokemon sun on my mac. I was able to play for a solid hour without any issues. Maybe try downloading the most recent nightly and see what happens 
However, the problem still persists with other games such as shin megami 4 and majora's mask idk why

**Comment by JayFoxRox** [comment]
All this talk without progress just lowers Signal to Noise ratio.

If you want it fixed, bisect it [google how to bisect / what bisecting does and do it]

**Comment by B3n30** [comment]
I spent some time, trying to find the issue. Here are my results:

- It's not a memory leak. The memory get freed as soon as the emuThreat is destroyed
- It's some memory OpenGL uses (probably for textures) and it never gets freed.
Using OpenGL Monitor Driver shows exactly that behavior:

<img width="914" alt="bildschirmfoto 2017-07-19 um 17 02 04" src="https://user-images.githubusercontent.com/26032316/28374307-a270481e-6ca4-11e7-90cb-bc927c26bc4c.png">
The first raising is while having the emu-window open. Closing the emu-window doesn't cause the textures to get destroyed. Only by starting the game again, the emu-threat (and therefore all textures) get destroyed.

**Comment by roberthharvey** [comment]
so ok, this was a problem in unity a while back, the fix there was a string that effectively forced non-viewed textures to be unloaded from memory stack. 

`Resources.UnloadUnusedAssets();`

probably the wrong language, probably not the fix, hope it still helps!

**Comment by pippo2931** [comment]
I don't see this anymore on High Sierra and latest citra.

**Comment by wwylele** [comment]
Fixed by #3281 as reported so closing. Feel free to reopen if it is not the case.
